nixos/modules/services/security/step-ca.nix

```
@@ -0,0 +1,125 @@
+          <literal>builtins.fromJSON</literal>.
+          <link xlink:href="https://smallstep.com/docs/step-cli/basic-crypto-operations#run-an-offline-x509-certificate-authority">This article</link>
+          may also be useful if you want to customize certain aspects of
+          certificate generation for your CA.
+        '';
+      };
+      intermediatePasswordFile = lib.mkOption {
+        type = with lib.types; nullOr path;
+        default = null;
+        example = "/run/keys/smallstep-password";
+        description = ''
+          Path to the file containing the password for the intermediate
+          certificate private key.
+
+          <warning>
+            <para>
+              Make sure to use a quoted absolute path instead of a path literal
+              to prevent it from being copied to the globally readable Nix
+              store.
+            </para>
+          </warning>
+        '';
+      };
+    };
+  };
+
+  config = lib.mkIf config.services.step-ca.enable (
+    let
+      configFile = settingsFormat.generate "ca.json"
+        (builtins.removeAttrs
+          (cfg.settings //
+            {
+              address = cfg.settings.address +
+                lib.optionalString (cfg.settings.port != null) ":" + toString cfg.settings.port;
+            })
+          [ "port" ]);
+    in
+    {
+      assertions =
+        [{
+          assertion = !lib.isStorePath cfg.intermediatePasswordFile;
+          message = ''
+            <option>services.step-ca.intermediatePasswordFile</option> points to
+            a file in the Nix store. You should use a quoted absolute path to
+            prevent this.
+          '';
+        }
+        {
+          assertion = !cfg.openFirewall || cfg.settings.port != null;
+          message = ''
+            When enabling opening the firewall a port must be specified explicitly.
+          '';
+        }];
+
+      systemd.packages = [ pkgs.step-ca ];
+
+      # configuration file indirection is needed to support reloading
+      environment.etc."smallstep-ca.json".source = configFile;
+
+      systemd.services."step-ca" = {
+        wantedBy = [ "multi-user.target" ];
+        restartTriggers = [ configFile ];
+        reloadIfChanged = true; # allow reloads for more service availability
+        unitConfig = {
+          ConditionFileNotEmpty = ""; # override upstream
+        };
+        serviceConfig = {
+          Environment = ""; # override upstream
+          WorkingDirectory = ""; # override upstream
+          ReadWriteDirectories = ""; # override upstream
+
+          LoadCredential = "intermediate_password:${cfg.intermediatePasswordFile}"; # DynamicUser means we don't know which user needs access to this file. Therefore let systemd handle this.
+
+          ExecStart = [
+            ""
+            ''${pkgs.step-ca}/bin/step-ca /etc/smallstep-ca.json --password-file ''${CREDENTIALS_DIRECTORY}/intermediate_password''
+          ]; # override upstream
```

Handle the case where `cfg.intermediatePasswordFile` is `null`.
```suggestion
          ExecStart = [
            "" # override upstream
            ("${pkgs.step-ca}/bin/step-ca /etc/ca.json"
              + optionalString (cfg.intermediatePasswordFile != null)
              " --password-file \${CREDENTIALS_DIRECTORY}/intermediate_password")
          ];
```

---

In my opinion we shouldn't allow setting this to null as then the private key would be word-readable in the NIx store. The current solution is already not optimal as the config file can contain oauth secrets afaik